src/orderbook.ts

```diff
@@ -103,6 +103,56 @@ export class Orderbook {
     return priceToCoverAmount(new Fraction(amount, 1), asks);
   }
 
+  /**
+   * Removes any overlapping bid/asks which could be matched in the current orderbook
+   * @return A new instance of the orderbook with no more overlapping orders.
+   */
+  reduced() {
+    const result = new Orderbook(this.baseToken, this.quoteToken);
+
+    const bids = Array.from(this.bids.values());
+    bids.sort(sortOffersDescending);
+    const asks = Array.from(this.asks.values());
+    asks.sort(sortOffersAscending);
+
+    const bid_iterator = bids.values();
+    const ask_iterator = asks.values();
+
+    let best_bid = bid_iterator.next();
+    let best_ask = ask_iterator.next();
+    while (
+      !(best_bid.done || best_ask.done) &&
+      !best_bid.value.price.lt(best_ask.value.price)
+    ) {
+      // We have an overlapping bid/ask. Subtract the smaller from the larger and remove the smaller
```

Comment:
Why doesn't the actual "matched" volume depend on the price at which they are matched?

Reply:
You are right that since our orders are all sellOrders, the price has an influence on the volume. The computation here is not exact. 
When creating the orderbook we collect all sellOrders as asks and then invert/add them together to get a full picture of bids and asks. There, the only price we have at hand is the limit price and we [multiply the volume with the limit price](https://github.com/gnosis/dex-contracts/blob/6d814fb02c46a559a7f235a20304797656853c05/src/orderbook.ts#L271). 
Since a matching price should be strictly "better" (that is higher than the limit) the available volume will be at least as large as the one computed here. 
Thus, we should find a lower bound of available volume which should be ok for estimating purposes.

Reply:
You are right that since our orders are all sellOrders, the price has an influence on the volume. The computation here is not exact. 
When creating the orderbook we collect all sellOrders as asks and then invert/add them together to get a full picture of bids and asks. There, the only price we have at hand is the limit price and we [multiply the volume with the limit price](https://github.com/gnosis/dex-contracts/blob/6d814fb02c46a559a7f235a20304797656853c05/src/orderbook.ts#L271). 
Since a matching price should be strictly "better" (that is higher than the limit) the available volume will be at least as large as the one computed here. 
Thus, we should find a lower bound of available volume which should be ok for estimating purposes.